backend/controllers/chatController.js

```diff
@@ -0,0 +1,154 @@
+const mongoose = require("mongoose");
+const path = require('path');
+const { v4: uuidv4 } = require('uuid');
+const { Chat, Message, Poll  } = require("../models/index");
+
+let connections = {};
+
+async function post_message(req, res) {
+    const { new_message, chat_id } = req.params;
+    update_clients_in_chat(new_message, chat_id);
+    res.send("User post message");
+}
+
+function update_clients_in_chat(new_update, chat_id) {
+    let clients = connections[chat_id];
+    clients.forEach(client => client.response.write(`data: ${JSON.stringify(new_update)}\n\n`));
+}
+
+async function subscribe_chat(req, res) {
+    try {
+        const { chat_id } = req.params;
+        const headers = {
+            'Content-Type': 'text/event-stream',
+            'Connection': 'keep-alive',
+            'Cache-Control': 'no-cache'
+        };
+        res.writeHead(200, headers);
+
+        const chat_buffer = await load_chat(chat_id, Date.now());
+        const data = `data: ${JSON.stringify(chat_buffer)}\n\n`;
+        res.write(data);
+
+        // Use a unique id for each client
+        const clientId = uuidv4(); 
+        // Add the client to the connections object
+        if (!connections[chat_id]) {
+            connections[chat_id] = [];
+        }
+        connections[chat_id].push({
+            id: clientId,
+            response: res
+        });
+
+        req.on("close", () => {
+            // Remove the client from the chat_id's connections
+            connections[chat_id] = connections[chat_id].filter(client => client.id !== clientId);
+            res.end(); // close the connection
+        })
+    } catch(err) {
+        throw new Error(`${err.message}`);
+    }
+}
+
+async function load_chat(chat_id, timestamp, buffer_size=10) {
+    try {
+        // validate chat_id (2 stesps)
+        const is_valid_id = mongoose.Types.ObjectId.isValid(chat_id);
+        if (!is_valid_id) {
+            throw new Error("Invalid chat id");
+        }
+
+        const existing_chat = await Chat.findById(chat_id);
+        if (!existing_chat) {
+            throw new Error("This chat doesn't exist");
+        }
+
+        // validate timestamp format
+        if (!is_valid_timestamp(timestamp)) {
+            throw new Error("Timestamp requested is invalid");
+        } 
+        // validate timestamp logic
+        if (new Date(timestamp) > Date.now()) {
+            throw new Error("Timestamp requested is in the future");
+        }
+
+        const chat = await Chat.findById({ _id: chat_id });
+        const messages = await load_message_buffer(chat.message, buffer_size, timestamp);
+        const polls = await load_poll_buffer(chat.polls, buffer_size, timestamp);
+        return sort_by_timestamp(messages, polls, buffer_size);
```

Comment:
Either we need to return the messages and polls as a json here so that view can format it, or we need to create another function to return it.

Reply:
I return an array so that we can keep the order of messages and polls in ascending order (latest -> oldest)

Reply:
yeah I get that, but we need a function that would send a valid response status code and the items in a json format so that the front end can use the information. Try doing res.status(200).send(sort_by_timestamp(messages, polls, buffer_size)) and check if the json is something that the frontend will be able to use to make the chat page

Reply:
So If we use JSON.parse() on the client-side, we can convert that string into an array to use [reference](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/JSON/parse#examples)